Build Failures in Candle

## Failure to build with CUDA/NCCL
<details>
<sumamary>Upstream update seems to be breaking NCCL/CUDA builds now:</summary>

```rust
90.39 error[E0599]: no method named `cuda_device` found for reference `&candle_core::CudaDevice` in the current scope
90.39    --> mistralrs-quant/src/distributed/mod.rs:214:48
90.39     |
90.39 214 |             let device = dev.as_cuda_device()?.cuda_device();
90.39     |                                                ^^^^^^^^^^^ method not found in `&candle_core::CudaDevice`
90.39 
90.43 error[E0599]: no method named `ordinal` found for struct `candle_core::CudaDevice` in the current scope
90.43    --> mistralrs-quant/src/distributed/mod.rs:493:44
90.43     |
90.43 493 | ...                   assert_eq!(dev.ordinal(), nccl_comm.rank());
90.43     |                                      ^^^^^^^ method not found in `candle_core::CudaDevice`
90.43 
90.44 error[E0599]: the method `w` exists for enum `Result<CudaSlice<half::bf16>, candle_core::Error>`, but its trait bounds were not satisfied
90.44    --> mistralrs-quant/src/distributed/mod.rs:495:84
90.44     |
90.44 495 | ...                   let mut dst = unsafe { dev.alloc::<bf16>(elem_count) }.w()?;
90.44     |                                                                              ^ method cannot be called on `Result<CudaSlice<half::bf16>, candle_core::Error>` due to unsatisfied trait bounds
90.44     |
90.44    ::: /root/.cargo/git/checkouts/candle-6740f55d69a3bf41/2f78d67/candle-core/src/error.rs:22:1
90.44     |
90.44 22  | pub enum Error {
90.44     | -------------- doesn't satisfy `candle_core::Error: Into<CudaError>`
90.44     |
90.44     = note: the following trait bounds were not satisfied:
90.44             `candle_core::Error: Into<CudaError>`
90.44             which is required by `Result<CudaSlice<half::bf16>, candle_core::Error>: WrapErr<CudaSlice<half::bf16>>`
90.44 
90.44 error[E0599]: no method named `ordinal` found for struct `candle_core::CudaDevice` in the current scope
90.44    --> mistralrs-quant/src/distributed/mod.rs:591:44
90.44     |
90.44 591 | ...                   assert_eq!(dev.ordinal(), nccl_comm.rank());
90.44     |                                      ^^^^^^^ method not found in `candle_core::CudaDevice`
90.44 
90.44 error[E0599]: the method `w` exists for enum `Result<CudaSlice<half::f16>, candle_core::Error>`, but its trait bounds were not satisfied
90.44    --> mistralrs-quant/src/distributed/mod.rs:508:83
90.44     |
90.44 508 | ...                   let mut dst = unsafe { dev.alloc::<f16>(elem_count) }.w()?;
90.44     |                                                                             ^ method cannot be called on `Result<CudaSlice<half::f16>, candle_core::Error>` due to unsatisfied trait bounds
90.44     |
90.44    ::: /root/.cargo/git/checkouts/candle-6740f55d69a3bf41/2f78d67/candle-core/src/error.rs:22:1
90.44     |
90.44 22  | pub enum Error {
90.44     | -------------- doesn't satisfy `candle_core::Error: Into<CudaError>`
90.44     |
90.44     = note: the following trait bounds were not satisfied:
90.44             `candle_core::Error: Into<CudaError>`
90.44             which is required by `Result<CudaSlice<half::f16>, candle_core::Error>: WrapErr<CudaSlice<half::f16>>`
90.44 
90.45 error[E0599]: the method `w` exists for enum `Result<CudaSlice<half::bf16>, candle_core::Error>`, but its trait bounds were not satisfied
90.45    --> mistralrs-quant/src/distributed/mod.rs:593:84
90.45     |
90.45 593 | ...                   let mut dst = unsafe { dev.alloc::<bf16>(elem_count) }.w()?;
90.45     |                                                                              ^ method cannot be called on `Result<CudaSlice<half::bf16>, candle_core::Error>` due to unsatisfied trait bounds
90.45     |
90.45    ::: /root/.cargo/git/checkouts/candle-6740f55d69a3bf41/2f78d67/candle-core/src/error.rs:22:1
90.45     |
90.45 22  | pub enum Error {
90.45     | -------------- doesn't satisfy `candle_core::Error: Into<CudaError>`
90.45     |
90.45     = note: the following trait bounds were not satisfied:
90.45             `candle_core::Error: Into<CudaError>`
90.45             which is required by `Result<CudaSlice<half::bf16>, candle_core::Error>: WrapErr<CudaSlice<half::bf16>>`
90.45 
90.45 error[E0599]: the method `w` exists for enum `Result<CudaSlice<f32>, candle_core::Error>`, but its trait bounds were not satisfied
90.45    --> mistralrs-quant/src/distributed/mod.rs:521:83
90.45     |
90.45 521 | ...                   let mut dst = unsafe { dev.alloc::<f32>(elem_count) }.w()?;
90.45     |                                                                             ^ method cannot be called on `Result<CudaSlice<f32>, candle_core::Error>` due to unsatisfied trait bounds
90.45     |
90.45    ::: /root/.cargo/git/checkouts/candle-6740f55d69a3bf41/2f78d67/candle-core/src/error.rs:22:1
90.45     |
90.45 22  | pub enum Error {
90.45     | -------------- doesn't satisfy `candle_core::Error: Into<CudaError>`
90.45     |
90.45     = note: the following trait bounds were not satisfied:
90.45             `candle_core::Error: Into<CudaError>`
90.45             which is required by `Result<CudaSlice<f32>, candle_core::Error>: WrapErr<CudaSlice<f32>>`
90.45 
90.46 error[E0599]: the method `w` exists for enum `Result<CudaSlice<half::f16>, candle_core::Error>`, but its trait bounds were not satisfied
90.46    --> mistralrs-quant/src/distributed/mod.rs:606:83
90.46     |
90.46 606 | ...                   let mut dst = unsafe { dev.alloc::<f16>(elem_count) }.w()?;
90.46     |                                                                             ^ method cannot be called on `Result<CudaSlice<half::f16>, candle_core::Error>` due to unsatisfied trait bounds
90.46     |
90.46    ::: /root/.cargo/git/checkouts/candle-6740f55d69a3bf41/2f78d67/candle-core/src/error.rs:22:1
90.46     |
90.46 22  | pub enum Error {
90.46     | -------------- doesn't satisfy `candle_core::Error: Into<CudaError>`
90.46     |
90.46     = note: the following trait bounds were not satisfied:
90.46             `candle_core::Error: Into<CudaError>`
90.46             which is required by `Result<CudaSlice<half::f16>, candle_core::Error>: WrapErr<CudaSlice<half::f16>>`
90.46 
90.47 error[E0599]: the method `w` exists for enum `Result<CudaSlice<f32>, candle_core::Error>`, but its trait bounds were not satisfied
90.47    --> mistralrs-quant/src/distributed/mod.rs:619:83
90.47     |
90.47 619 | ...                   let mut dst = unsafe { dev.alloc::<f32>(elem_count) }.w()?;
90.47     |                                                                             ^ method cannot be called on `Result<CudaSlice<f32>, candle_core::Error>` due to unsatisfied trait bounds
90.47     |
90.47    ::: /root/.cargo/git/checkouts/candle-6740f55d69a3bf41/2f78d67/candle-core/src/error.rs:22:1
90.47     |
90.47 22  | pub enum Error {
90.47     | -------------- doesn't satisfy `candle_core::Error: Into<CudaError>`
90.47     |
90.47     = note: the following trait bounds were not satisfied:
90.47             `candle_core::Error: Into<CudaError>`
90.47             which is required by `Result<CudaSlice<f32>, candle_core::Error>: WrapErr<CudaSlice<f32>>`
90.47 
90.51 warning: unused import: `driver::DeviceSlice`
90.51    --> mistralrs-quant/src/distributed/mod.rs:478:26
90.51     |
90.51 478 |             use cudarc::{driver::DeviceSlice, nccl::ReduceOp};
90.51     |                          ^^^^^^^^^^^^^^^^^^^
90.51     |
90.51     = note: `#[warn(unused_imports)]` on by default
90.51 
90.51 warning: unused import: `cudarc::driver::DeviceSlice`
90.51    --> mistralrs-quant/src/distributed/mod.rs:572:17
90.51     |
90.51 572 |             use cudarc::driver::DeviceSlice;
90.51     |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^
90.51 
91.02 For more information about this error, try `rustc --explain E0599`.
91.04 warning: `mistralrs-quant` (lib) generated 2 warnings
91.04 error: could not compile `mistralrs-quant` (lib) due to 9 previous errors; 2 warnings emitted
91.04 warning: build failed, waiting for other jobs to finish...
```
</details>

## Latest commit or version
Current master with #1468 (targeting 12.9.1)

--- a/mistralrs-quant/src/distributed/mod.rs
+++ b/mistralrs-quant/src/distributed/mod.rs
@@ -211,14 +211,14 @@ mod nccl {
                     world_size
                 );
             }
-            let device = dev.as_cuda_device()?.cuda_device();
-            assert_eq!(rank, device.ordinal());
+            let stream = dev.as_cuda_device()?.cuda_stream();
+            assert_eq!(rank, stream.context().ordinal());
             let nccl_id = match id {
                 super::Id::Nccl(id) => id,
                 _ => panic!("Expected NCCL Id variant"),
             };
             Ok(Self {
-                comm: cudarc::nccl::Comm::from_rank(device, rank, world_size, nccl_id)
+                comm: cudarc::nccl::Comm::from_rank(stream, rank, world_size, nccl_id)
                     .map_err(|e| e.0)
                     .expect("Failed to create `Comm`, error code"),
             })
@@ -440,8 +440,8 @@ mod nccl_ops {
     use std::{fmt::Debug, sync::Arc};
 
     use candle_core::{
-        backend::BackendStorage, cuda::cudarc, cuda_backend::WrapErr, CpuStorage, CustomOp1, DType,
-        Layout, Result, Shape, Tensor,
+        backend::BackendStorage, cuda::cudarc, CpuStorage, CustomOp1, DType, Layout, Result, Shape,
+        Tensor,
     };
 
     #[derive(Clone, Debug)]
@@ -475,7 +475,7 @@ mod nccl_ops {
             s: &candle_core::CudaStorage,
             l: &Layout,
         ) -> Result<(candle_core::CudaStorage, Shape)> {
-            use cudarc::{driver::DeviceSlice, nccl::ReduceOp};
+            use cudarc::nccl::ReduceOp;
             use half::{bf16, f16};
 
             let elem_count = l.shape().elem_count();
@@ -490,9 +490,9 @@ mod nccl_ops {
                                 Some((0, l)) if l == s.len() => s,
                                 Some(_) | None => candle_core::bail!("input has to be contiguous"),
                             };
-                            assert_eq!(dev.ordinal(), nccl_comm.rank());
                             assert!(elem_count > 0);
-                            let mut dst = unsafe { dev.alloc::<bf16>(elem_count) }.w()?;
+                            assert_eq!(dev.cuda_stream().context().ordinal(), nccl_comm.rank());
+                            let mut dst = unsafe { dev.alloc::<bf16>(elem_count) }?;
                             nccl_comm
                                 .inner()
                                 .all_reduce(s, &mut dst, &ReduceOp::Sum)
@@ -505,7 +505,9 @@ mod nccl_ops {
                                 Some((0, l)) if l == s.len() => s,
                                 Some(_) | None => candle_core::bail!("input has to be contiguous"),
                             };
-                            let mut dst = unsafe { dev.alloc::<f16>(elem_count) }.w()?;
+                            assert!(elem_count > 0);
+                            assert_eq!(dev.cuda_stream().context().ordinal(), nccl_comm.rank());
+                            let mut dst = unsafe { dev.alloc::<f16>(elem_count) }?;
                             nccl_comm
                                 .inner()
                                 .all_reduce(s, &mut dst, &ReduceOp::Sum)
@@ -518,7 +520,9 @@ mod nccl_ops {
                                 Some((0, l)) if l == s.len() => s,
                                 Some(_) | None => candle_core::bail!("input has to be contiguous"),
                             };
-                            let mut dst = unsafe { dev.alloc::<f32>(elem_count) }.w()?;
+                            assert!(elem_count > 0);
+                            assert_eq!(dev.cuda_stream().context().ordinal(), nccl_comm.rank());
+                            let mut dst = unsafe { dev.alloc::<f32>(elem_count) }?;
                             nccl_comm
                                 .inner()
                                 .all_reduce(s, &mut dst, &ReduceOp::Sum)
@@ -569,7 +573,6 @@ mod nccl_ops {
             s: &candle_core::CudaStorage,
             l: &Layout,
         ) -> Result<(candle_core::CudaStorage, Shape)> {
-            use cudarc::driver::DeviceSlice;
             use half::{bf16, f16};
 
             let mut out_shape = l.shape().dims().to_vec();
@@ -588,9 +591,9 @@ mod nccl_ops {
                                 Some((0, l)) if l == s.len() => s,
                                 Some(_) | None => candle_core::bail!("input has to be contiguous"),
                             };
-                            assert_eq!(dev.ordinal(), nccl_comm.rank());
                             assert!(elem_count > 0);
-                            let mut dst = unsafe { dev.alloc::<bf16>(elem_count) }.w()?;
+                            assert_eq!(dev.cuda_stream().context().ordinal(), nccl_comm.rank());
+                            let mut dst = unsafe { dev.alloc::<bf16>(elem_count) }?;
                             nccl_comm
                                 .inner()
                                 .all_gather(s, &mut dst)
@@ -603,7 +606,9 @@ mod nccl_ops {
                                 Some((0, l)) if l == s.len() => s,
                                 Some(_) | None => candle_core::bail!("input has to be contiguous"),
                             };
-                            let mut dst = unsafe { dev.alloc::<f16>(elem_count) }.w()?;
+                            assert!(elem_count > 0);
+                            assert_eq!(dev.cuda_stream().context().ordinal(), nccl_comm.rank());
+                            let mut dst = unsafe { dev.alloc::<f16>(elem_count) }?;
                             nccl_comm
                                 .inner()
                                 .all_gather(s, &mut dst)
@@ -616,7 +621,9 @@ mod nccl_ops {
                                 Some((0, l)) if l == s.len() => s,
                                 Some(_) | None => candle_core::bail!("input has to be contiguous"),
                             };
-                            let mut dst = unsafe { dev.alloc::<f32>(elem_count) }.w()?;
+                            assert!(elem_count > 0);
+                            assert_eq!(dev.cuda_stream().context().ordinal(), nccl_comm.rank());
+                            let mut dst = unsafe { dev.alloc::<f32>(elem_count) }?;
                             nccl_comm
                                 .inner()
                                 .all_gather(s, &mut dst)

